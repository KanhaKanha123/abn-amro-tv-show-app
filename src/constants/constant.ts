const AppRoutes = {
  dashboard: '/',
  showDetails: '/show-details',
}

const ItemCountPerPage = {
  itemPerGenre: 5
}

const ApiEndPoint = {
  apiShows: 'https://api.tvmaze.com/shows'
}

export const RoutesConfig = {
  ...AppRoutes,
  ...ApiEndPoint,
  ...ItemCountPerPage,
};
