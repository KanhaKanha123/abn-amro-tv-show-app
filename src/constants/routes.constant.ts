enum AppRoutes {
  dashboard = '/',
  showDetails = '/show-details',

}
enum ItemCountPerPage{
  itemPerGenre=6
}
enum ApiRoutes {
  apiShows = 'https://api.tvmaze.com/shows',
  apiSearchShows = 'https://api.tvmaze.com/search/shows',
}

export const RoutesConfig = {
  ...AppRoutes,
  ...ApiRoutes,
  ...ItemCountPerPage,
};
