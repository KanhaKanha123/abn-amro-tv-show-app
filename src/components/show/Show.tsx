import { Card } from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import Rating from '../ratings/Rating';
import { RoutesConfig } from '../../constants/constant';
import { ShowType } from '../../types/types';

const Show = ({ item }: any) => {
    const navigate = useNavigate();

    const navigateToShowDetails = (showDetails: ShowType): void => {
        navigate(RoutesConfig.showDetails, { state: { showDetails } });
    };

    return <main aria-label='TV show detail cart' data-testid="show-details-container" className="show-details">
        <Card data-testid="show-details-card" className='card-container' onClick={() => navigateToShowDetails(item)}>
            <Card.Img aria-label="Image of show" data-testid="show-details-image" variant='top' src={item.image} alt="show image here"></Card.Img>
            <Card.Subtitle aria-label="name of show" data-testid="show-details-name" className='show-name'>{item.name}</Card.Subtitle>
            <Card.Subtitle className='card-subtitle'>
                <Card.Text aria-label="genres of show" data-testid="show-details-genres" className='small-text'>{item.genres}</Card.Text>
                <Rating aria-label="Average rating of show" rating={(item.rating)}></Rating>
            </Card.Subtitle>
        </Card>
    </main>
}

export default Show;