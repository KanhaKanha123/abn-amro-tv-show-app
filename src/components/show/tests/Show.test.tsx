import Show from '../Show';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { ShowComponentMockProps } from './fixture/test-data';
import * as router from 'react-router'
import { RoutesConfig } from '../../../constants/constant';

const navigate = jest.fn()

describe('Show component test', () => {
    beforeEach(() => {
        jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate)
        render(<Show item={ShowComponentMockProps} />);
    });

    test("Show component load correctly", () => {
        const wrapper = screen.getByTestId("show-details-container");
        const image = screen.getByTestId("show-details-image");

        // check element exist
        expect(wrapper).toBeInTheDocument();
        expect(image).toBeInTheDocument();
    });

    test("Show component should recieve correct props", () => {
        const showName = screen.getByTestId("show-details-name");
        const showGenres = screen.getByTestId("show-details-genres");

        //check tags have coorect values
        expect(showName).toHaveTextContent('Under')
        expect(showGenres).toHaveTextContent('Drama')
    });

    test("navigateToShowDetails should be called", async () => {
        const element = await renderer.create(<Show item={ShowComponentMockProps} />);
        const showCard = screen.getByTestId("show-details-card");
        fireEvent.click(showCard);
        expect(navigate).toHaveBeenCalledWith(RoutesConfig.showDetails, { "state": { "showDetails": { "fullName": "Under", "genres": "Drama", "id": 1, "image": "https://static.tvmaze.com/uploads/images/medium_portrait/81/202627.jpg", "name": "Under", "rating": 6.5, "summary": "Under the Dome" } } });
    });

    test("Show component should match snapshot", async () => {
        const element = await renderer.create(<Show item={ShowComponentMockProps} />);
        expect(element.toJSON()).toMatchSnapshot();
    });

});
