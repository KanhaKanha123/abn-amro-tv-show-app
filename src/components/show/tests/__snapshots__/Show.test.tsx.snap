// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show component test Show component should match snapshot 1`] = `
<main
  aria-label="TV show detail cart"
  className="show-details"
  data-testid="show-details-container"
>
  <div
    className="card-container card"
    data-testid="show-details-card"
    onClick={[Function]}
  >
    <img
      alt="show image here"
      aria-label="Image of show"
      className="card-img-top"
      data-testid="show-details-image"
      src="https://static.tvmaze.com/uploads/images/medium_portrait/81/202627.jpg"
    />
    <div
      aria-label="name of show"
      className="show-name card-subtitle h6"
      data-testid="show-details-name"
    >
      Under
    </div>
    <div
      className="card-subtitle card-subtitle h6"
    >
      <p
        aria-label="genres of show"
        className="small-text card-text"
        data-testid="show-details-genres"
      >
        Drama
      </p>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          data-testid="rating-half-star7"
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.354 5.119 7.538.792A.516.516 0 0 1 8 .5c.183 0 .366.097.465.292l2.184 4.327 4.898.696A.537.537 0 0 1 16 6.32a.548.548 0 0 1-.17.445l-3.523 3.356.83 4.73c.078.443-.36.79-.746.592L8 13.187l-4.389 2.256a.52.52 0 0 1-.146.05c-.342.06-.668-.254-.6-.642l.83-4.73L.173 6.765a.55.55 0 0 1-.172-.403.58.58 0 0 1 .085-.302.513.513 0 0 1 .37-.245l4.898-.696zM8 12.027a.5.5 0 0 1 .232.056l3.686 1.894-.694-3.957a.565.565 0 0 1 .162-.505l2.907-2.77-4.052-.576a.525.525 0 0 1-.393-.288L8.001 2.223 8 2.226v9.8z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
          />
        </svg>
      </span>
      <span
        aria-label="rating star image"
      >
        <svg
          fill="currentColor"
          fontSize="15px"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
          />
        </svg>
      </span>
    </div>
  </div>
</main>
`;
